include/dlaf/eigensolver/tridiag_solver/merge.h

```diff
@@ -0,0 +1,923 @@
+                                                      T* d_ptr, T* z_ptr, ColType* c_ptr) {
+  std::vector<GivensRotation<T>> rots;
+  rots.reserve(to_sizet(len));
+
+  SizeType i1 = 0;  // index of 1st element in the Givens rotation
+  // Iterate over the indices of the sorted elements in pair (i1, i2) where i1 < i2 for every iteration
+  for (SizeType i2 = 1; i2 < len; ++i2) {
+    SizeType i1s = i_ptr[i1];
+    SizeType i2s = i_ptr[i2];
+    T& d1 = d_ptr[i1s];
+    T& d2 = d_ptr[i2s];
+    T& z1 = z_ptr[i1s];
+    T& z2 = z_ptr[i2s];
+    ColType& c1 = c_ptr[i1s];
+    ColType& c2 = c_ptr[i2s];
+
+    // if z1 nearly zero deflate the element and move i1 forward to i2
+    if (std::abs(rho * z1) <= tol) {
+      c1 = ColType::Deflated;
+      i1 = i2;
+      continue;
+    }
+
+    // Deflate the second element if z2 nearly zero
+    if (std::abs(rho * z2) <= tol) {
+      c2 = ColType::Deflated;
+      continue;
+    }
+
+    // Given's deflation condition is the same as the one used in LAPACK's stedc implementation [1].
+    //
+    // [1] LAPACK 3.10.0, file dlaed2.f, line 393
+    T r = std::sqrt(z1 * z1 + z2 * z2);
+    T c = z1 / r;
+    T s = -z2 / r;
+
+    // If d1 is not nearly equal to d2, move i1 forward to i2
+    if (std::abs(c * s * (d2 - d1)) > tol) {
+      i1 = i2;
+      continue;
+    }
+
+    // When d1 is nearly equal to d2 apply Givens rotation
+    z1 = r;
+    z2 = 0;
+    T tmp = d1 * s * s + d2 * c * c;
+    d1 = d1 * c * c + d2 * s * s;
+    d2 = tmp;
+
+    rots.push_back(GivensRotation<T>{i1s, i2s, c, s});
+    //  Set the the `i1` column as "Dense" if the `i2` column has opposite non-zero structure (i.e if
+    //  one comes from Q1 and the other from Q2 or vice-versa)
+    if ((c1 == ColType::UpperHalf && c2 == ColType::LowerHalf) ||
+        (c1 == ColType::LowerHalf && c2 == ColType::UpperHalf)) {
+      c1 = ColType::Dense;
+    }
+    c2 = ColType::Deflated;
+  }
+
+  return rots;
+}
+
+template <class T>
+pika::future<std::vector<GivensRotation<T>>> applyDeflation(
+    SizeType i_begin, SizeType i_end, pika::shared_future<T> rho_fut, pika::shared_future<T> tol_fut,
+    Matrix<const SizeType, Device::CPU>& index, Matrix<T, Device::CPU>& d, Matrix<T, Device::CPU>& z,
+    Matrix<ColType, Device::CPU>& c) {
+  SizeType n = problemSize(i_begin, i_end, index.distribution());
+
+  auto deflate_fn = [n](auto rho_fut, auto tol_fut, auto index_tiles, auto d_tiles, auto z_tiles,
+                        auto c_tiles) {
+    TileElementIndex zero_idx(0, 0);
+    const SizeType* i_ptr = index_tiles[0].get().ptr(zero_idx);
+    // save in variable avoid releasing the tile too soon
+    auto d_tile = d_tiles[0].get();
+    auto z_tile = z_tiles[0].get();
+    auto c_tile = c_tiles[0].get();
+    T* d_ptr = d_tile.ptr(zero_idx);
+    T* z_ptr = z_tile.ptr(zero_idx);
+    ColType* c_ptr = c_tile.ptr(zero_idx);
+
+    return applyDeflationToArrays(rho_fut.get(), tol_fut.get(), n, i_ptr, d_ptr, z_ptr, c_ptr);
```

Comment:
Could the tiles be passed by reference to `applyDeflationToArrays` instead of passing pointers?

Reply:
Ah, I think I've answered my own question... You do it for the offsets, correct? It would be really nice to have a (sub)view into a tile (or do we already have some helpers for that? @albestro?).